### ISSUE TITLE
为啥不能用model.summary()查看模型结构呢

### ISSUE THREAD

**Issue by cl886699** [cmd23333/Faster-rcnn-tensorflow2.x#2]

**Comment by cmd23333** [comment]
您好，我不知道你是什么时候调用model.summary()的，因为通常我们需要**build模型**（即确认了input_shape）**之后，才可以显示summary**。我的程序里并没有Build模型这一步，不过通过调用 model.__call__(input) (这相当于model.fit()）就相当于是build了。
所以你不妨尝试一下，在model(img, bbox, label, scale)之后再调用model.summary()

如果还是不行，那就可能是**因为我的模型是完全自定义的**，需要传入额外的bbox/label位置信息，这**导致了不能.summary()**吧。

**Comment by cl886699** [comment]
谢谢你的回复，在model(img, bbox, label, scale)之后，还是不行，说是没有build。不能summary看看模型，总感觉心里不踏实


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "cmd23333/Faster-rcnn-tensorflow2.x"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月14日(星期二) 上午10:42
收件人:&nbsp;"cmd23333/Faster-rcnn-tensorflow2.x"<Faster-rcnn-tensorflow2.x@noreply.github.com&gt;;
抄送:&nbsp;"开开心心每一天"<88659445@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [cmd23333/Faster-rcnn-tensorflow2.x] 为啥不能用model.summary()查看模型结构呢 (#2)





 
您好，我不知道你是什么时候调用model.summary()的，因为通常我们需要build模型（即确认了input_shape）之后，才可以显示summary。我的程序里并没有Build模型这一步，不过通过调用 model.call(input) (这相当于model.fit()）就相当于是build了。
 所以你不妨尝试一下，在model(img, bbox, label, scale)之后再调用model.summary()
 
如果还是不行，那就可能是因为我的模型是完全自定义的，需要传入额外的bbox/label位置信息，这**导致了不能.summary()**吧。

**Comment by cmd23333** [comment]
我帮你找了一下，看看下面这个连接写的有没有帮助
https://www.cnblogs.com/yaos/articles/12739923.html

**Comment by cl886699** [comment]
谢谢，你太棒了





&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "cmd23333/Faster-rcnn-tensorflow2.x"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月15日(星期三) 中午11:20
收件人:&nbsp;"cmd23333/Faster-rcnn-tensorflow2.x"<Faster-rcnn-tensorflow2.x@noreply.github.com&gt;;
抄送:&nbsp;"开开心心每一天"<88659445@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [cmd23333/Faster-rcnn-tensorflow2.x] 为啥不能用model.summary()查看模型结构呢 (#2)





 
我帮你找了一下，看看下面这个连接写的有没有帮助
 https://www.cnblogs.com/yaos/articles/12739923.html

**Comment by cl886699** [comment]
我试了下，按照上面的方式想要打印模型结构，还需要把模型定义的__call__改成call，这里为什么要定义成__call__呢
&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "cmd23333/Faster-rcnn-tensorflow2.x"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月15日(星期三) 中午11:20
收件人:&nbsp;"cmd23333/Faster-rcnn-tensorflow2.x"<Faster-rcnn-tensorflow2.x@noreply.github.com&gt;;
抄送:&nbsp;"开开心心每一天"<88659445@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [cmd23333/Faster-rcnn-tensorflow2.x] 为啥不能用model.summary()查看模型结构呢 (#2)





 
我帮你找了一下，看看下面这个连接写的有没有帮助
 https://www.cnblogs.com/yaos/articles/12739923.html

**Comment by cmd23333** [comment]
__call__和call应该是一样的哈，就是实现了前向传播的功能。看下图，__call__和call指向是一样的。
<img width="868" alt="微信截图_20200716140318" src="https://user-images.githubusercontent.com/59002739/87633017-2f785780-c76d-11ea-9e4a-293a170fbb14.png">
summary的话 我有空了再看下

**Comment by cl886699** [comment]
这样显示的调用确实是一样的，但是对于用__call__定义的，是不能直接用model.build的，这个函数默认是去调用call。所以既然都可以，为啥不直接用call呢

**Comment by cmd23333** [comment]
我确认了一下，官网的API也是让我们定义call。。是我记错了 0 0. 那改成call之后可以build吗？

**Comment by cl886699** [comment]
有些可以build，整体还是不能build




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "cmd23333/Faster-rcnn-tensorflow2.x"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月17日(星期五) 上午10:51
收件人:&nbsp;"cmd23333/Faster-rcnn-tensorflow2.x"<Faster-rcnn-tensorflow2.x@noreply.github.com&gt;;
抄送:&nbsp;"开开心心每一天"<88659445@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [cmd23333/Faster-rcnn-tensorflow2.x] 为啥不能用model.summary()查看模型结构呢 (#2)





 
我确认了一下，官网的API也是让我们定义call。。是我记错了 0 0. 那改成call之后可以build吗？
